streamingccl: support canceling an ingestion job

gh-casper opened the pull request:
This PR supports the intended behavior when canceling
    an ingestion job: canceling the producer job which releases the
    protected timestamp for the source tenant. The
    destination tenant key ranges are left for
    the user to clean up with crdb_internal.destroy_tenant.

    This PR also expands crdb_internal.complete_stream_replication
    to support canceling a producer job to take 2nd argument
    'successfulIngestion' to support complete replication stream
    even when cutover doesn't happen, e.g., ingestion gets canceled
    and a revert happens, this cancels the producer job

    Release note (sql change): expand crdb_internal.complete_stream_replication
    to take successfulIngestion argument, which indicates if this
    stream ingestion finished successfully.

Closes: https://github.com/cockroachdb/cockroach/issues/57409

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/83310)
<!-- Reviewable:end -->

gh-casper reviewed (commented):
Please only review the top two commits.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/83310#-:-N5JgRnwE_MIiJhPvgyH:b-ki8qxn)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @miretskiy, @samiskin, and @stevendanna)



<!-- Sent from Reviewable.io -->

miretskiy commented on pkg/ccl/streamingccl/streamingest/stream_ingestion_job.go:
```diff
@@ -299,7 +299,59 @@ func revertToCutoverTimestamp(
+		return err
+	}
+
+	defer func() {
```
I would move this function into a helper.

miretskiy commented on pkg/ccl/streamingccl/streamingest/stream_ingestion_job.go:
```diff
@@ -299,7 +299,59 @@ func revertToCutoverTimestamp(
+		}
+
+		// TODO(casper): handle timeout when we have very large tenant with
+		// saving checkpoints and retrying.
```
I'm not sure about timeout (we haven't set it have we? And the context doesn't have a deadline set either); but I do think we need to add checks to ensure the requirements specified by the ClearRange request: namely, we need to ensure that the spans are offline.
https://github.com/cockroachdb/cockroach/blob/master/pkg/roachpb/api.proto#L384-L384

miretskiy commented on pkg/ccl/streamingccl/streamingest/stream_ingestion_job.go:
```diff
@@ -299,7 +299,59 @@ func revertToCutoverTimestamp(
+		spans = spans[:0]
+		for _, raw := range b.RawResponse().Responses {
+			r := raw.GetClearRange()
+			if r.ResumeSpan != nil {
```
I'm not sure GetClearRange returns resume spans.  I mean it has a header, and the field; but I'm not sure it ever returns resume spans.  My hunch is not; but would be nice to double check with @erikgrinaker

erikgrinaker replied:
No, `ClearRange` doesn't return resume spans.

miretskiy commented on pkg/ccl/streamingccl/streamingest/stream_replication_e2e_test.go:
```diff
@@ -358,6 +367,7 @@ func TestTenantStreamingPauseResumeIngestion(t *testing.T) {
 
 	// Pause ingestion.
 	c.destSysSQL.Exec(t, fmt.Sprintf("PAUSE JOB %d", ingestionJobID))
+	fmt.Println("after pausing job")
```
stray fmt.Println

miretskiy commented on pkg/ccl/streamingccl/streamingest/stream_replication_e2e_test.go:
```diff
@@ -428,3 +438,55 @@ func TestTenantStreamingPauseOnError(t *testing.T) {
+		destTenantPrefix := keys.MakeTenantPrefix(c.args.destTenantID)
+		rows, err := c.destKvDB.Scan(ctx, destTenantPrefix, destTenantPrefix.PrefixEnd(), 10)
+		require.NoError(t, err)
+		require.Empty(t, rows)
```
nice.

miretskiy commented on pkg/sql/sem/builtins/replication_builtins.go:
```diff
@@ -300,6 +300,7 @@ var replicationBuiltins = map[string]builtinDefinition{
 		tree.Overload{
 			Types: tree.ArgTypes{
 				{"stream_id", types.Int},
+				{"ingestion_cutover", types.Bool},
```
Do we ever not want to do this? I.e. do we imagine this function, which intended to be invoked from destination cluster, will ever want to invoke this set to false?

Is the intention to store some information about cutover event in the src cluster? If so, perhaps having a timestamp makes more sense.  Or even some sort of proto message indicating the reason for cutover, timestamp, etc..?

stevendanna reviewed (commented):
Overall the last two commits LGTM modulo the existing comments from @miretskiy.

gh-casper reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/83310)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @miretskiy, @samiskin, and @stevendanna)

___
*[`pkg/ccl/streamingccl/streamingest/stream_ingestion_job.go` line 313 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/83310#-N5JgMtnAMPrQiJzGBmq-r4-313:-N5d23586J9Dh8P-NVmB:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/41665f389da0e36fde3d049cf12c413afbd06750/pkg/ccl/streamingccl/streamingest/stream_ingestion_job.go#L313)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

I would move this function into a helper.  
</blockquote></details>

Done.
___
*[`pkg/ccl/streamingccl/streamingest/stream_ingestion_job.go` line 341 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/83310#-N5JgMtnAMPrQiJzGBmq-r4-341:-N5cwYu23KK4e3KdpTEd:b-5qenr7) ([raw file](https://github.com/cockroachdb/cockroach/blob/41665f389da0e36fde3d049cf12c413afbd06750/pkg/ccl/streamingccl/streamingest/stream_ingestion_job.go#L341)):*
<details><summary><i>Previously, stevendanna (Steven Danna) wrote…</i></summary><blockquote>

I believe we have some planned work to ensure that the tenant is offline throughout the ingestion process.
</blockquote></details>

Yes, this is the work that Shiranka is working on. Added a TODO here.
___
*[`pkg/ccl/streamingccl/streamingest/stream_ingestion_job.go` line 345 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/83310#-N5JgMtnAMPrQiJzGBmq-r4-345:-N5d0HGgCQbUFhOWTWIT:b-5fkguw) ([raw file](https://github.com/cockroachdb/cockroach/blob/41665f389da0e36fde3d049cf12c413afbd06750/pkg/ccl/streamingccl/streamingest/stream_ingestion_job.go#L345)):*
<details><summary><i>Previously, erikgrinaker (Erik Grinaker) wrote…</i></summary><blockquote>

No, `ClearRange` doesn't return resume spans.
</blockquote></details>

Okay. Removed. Added a TODO to deal with large tenant.
___
*[`pkg/ccl/streamingccl/streamproducer/stream_lifetime.go` line 162 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/83310#-N5JgMtnAMPrQiJzGBmx-r3-162:-N5d0U2qCfQh6MKdMqw4:bb6un1j) ([raw file](https://github.com/cockroachdb/cockroach/blob/9a1b9bff6b7a6fb68d6103040dff622f710b3f1b/pkg/ccl/streamingccl/streamproducer/stream_lifetime.go#L162)):*
<details><summary><i>Previously, stevendanna (Steven Danna) wrote…</i></summary><blockquote>

Should we use errors.Is here?
</blockquote></details>

Done
___
*[`pkg/sql/sem/builtins/replication_builtins.go` line 303 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/83310#-N5JgMtnAMPrQiJzGBn--r3-303:-N5d0ZXj9FhE5IV8DZaI:b1rq8to) ([raw file](https://github.com/cockroachdb/cockroach/blob/9a1b9bff6b7a6fb68d6103040dff622f710b3f1b/pkg/sql/sem/builtins/replication_builtins.go#L303)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

Do we ever not want to do this? I.e. do we imagine this function, which intended to be invoked from destination cluster, will ever want to invoke this set to false?

Is the intention to store some information about cutover event in the src cluster? If so, perhaps having a timestamp makes more sense.  Or even some sort of proto message indicating the reason for cutover, timestamp, etc..?
</blockquote></details>

This bool basically tells whether the stream ingestion ends with cutover or not. If it ends with cutover, then this is a successful ingestion, we mark producer job succeeded. If no, then we don't need to track this stream replication anymore, so we just cancel the producer job. I think logically this makes sense.

I made two commits on this PR, first commit introducing this flag, purpose being to support canceling the producer job when the ingestion job is being canceled and tenant key range gets reverted.
___
*[`pkg/ccl/streamingccl/streamingest/stream_replication_e2e_test.go` line 370 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/83310#-N5JgMtnAMPrQiJzGBmt-r4-370:-N5c8fFV80oAuqd8EXPl:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/41665f389da0e36fde3d049cf12c413afbd06750/pkg/ccl/streamingccl/streamingest/stream_replication_e2e_test.go#L370)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

stray fmt.Println
</blockquote></details>

Done.



_Code quote:_
```Go
fter
```
___
*[`pkg/ccl/streamingccl/streamingest/stream_replication_e2e_test.go` line 482 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/83310#-N5JgMtnAMPrQiJzGBmt-r4-482:-N5blkGY0pmmt3ATWPBh:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/41665f389da0e36fde3d049cf12c413afbd06750/pkg/ccl/streamingccl/streamingest/stream_replication_e2e_test.go#L482)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

nice.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

miretskiy commented on pkg/ccl/streamingccl/streamingest/stream_ingestion_job.go:
```diff
@@ -304,8 +320,29 @@ func revertToCutoverTimestamp(
+	// ingestion anymore.
+	defer cancelProducerJob(ctx,
+		streaming.StreamID(details.StreamID),
+		streamingccl.StreamAddress(details.StreamAddress))
```
you are missing `()` after defer, I believe.
Also, did any tests fail?

miretskiy replied:
Do we want to cancel if db.Run below failed?

miretskiy commented on pkg/ccl/streamingccl/streamingest/stream_ingestion_job.go:
```diff
@@ -304,8 +320,29 @@ func revertToCutoverTimestamp(
+	// TODO(casper): deal with very large tenant.
+	// TODO(casper): ensure the tenant is offline while we revert it.
+	var b kv.Batch
+	b.AddRawRequest(&roachpb.ClearRangeRequest{
```
Don't we want to to do this only if job was cancelled? And not e.g. because some error happen?

gh-casper reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/83310)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @miretskiy, @samiskin, and @stevendanna)

___
*[`pkg/ccl/streamingccl/streamingest/stream_ingestion_job.go` line 334 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/83310#-N5JgMtnAMPrQiJzGBmq-r9-334:-N5fDVm07jBZK0MVofFD:bklt9bm) ([raw file](https://github.com/cockroachdb/cockroach/blob/1ca080fb9cdaf34375ae7347e99d6ee4d45284ba/pkg/ccl/streamingccl/streamingest/stream_ingestion_job.go#L334)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

Do we want to cancel if db.Run below failed?
</blockquote></details>

No. cancelProducerJob is a function defined outside of this function body. Adding `() ` after defer doesn't make it compile. 

Yes, if db.Run below failed, 	we still want to cancel the producer job because the source job's
protected timestamp is no longer needed as this ingestion job won't resume ingestion anymore (already in cancel-requested status). This makes me consider doing `cancelProducerJob` before `ClearRange`.
___
*[`pkg/ccl/streamingccl/streamingest/stream_ingestion_job.go` line 339 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/83310#-N5JgMtnAMPrQiJzGBmq-r9-339:-N5fEzynE4AvihjgJG5C:b-msd8dm) ([raw file](https://github.com/cockroachdb/cockroach/blob/1ca080fb9cdaf34375ae7347e99d6ee4d45284ba/pkg/ccl/streamingccl/streamingest/stream_ingestion_job.go#L339)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

Don't we want to to do this only if job was cancelled? And not e.g. because some error happen?
</blockquote></details>

We will pause on error, so only `CANCEL JOB` can trigger the `ClearRange`.


<!-- Sent from Reviewable.io -->

miretskiy commented on pkg/ccl/streamingccl/streamingest/stream_ingestion_job.go:
```diff
@@ -105,39 +109,67 @@ type streamIngestionResumer struct {
+		// Make sure the producer job is active before start the stream replication.
+		var status streampb.StreamReplicationStatus
+		for r := retry.Start(ro); r.Next(); {
+			status, err = client.Heartbeat(ctx, streamID, startTime)
```
I thought that the ingestion processor was supposed to heartbeat src cluster periodically; I only see this happen here... where do we periodically ping src cluster?

miretskiy commented on pkg/ccl/streamingccl/streamingest/stream_ingestion_job.go:
```diff
@@ -148,24 +180,32 @@ func ingest(
 		streamIngestionSpecs, streamIngestionFrontierSpec, err := distStreamIngestionPlanSpecs(
-			streamAddress, topology, sqlInstanceIDs, initialHighWater, ingestionJobID, streamID, oldTenantID, newTenantID)
+			streamAddress, topology, sqlInstanceIDs, progress.GetStreamIngest().StartTime,
+			ingestionJob.ID(), streamID, details.TenantID, details.NewTenantID)
```
I don't see in this code how do we ever retry (i.e. hit the `for r := retry.Start(ro); r.Next(); {` line).  It seems that we bubble up every error encountered in this function.

distStreamIngestion for example could return an error for any number of legitimate reasons (node crashed; or perhaps restarted, etc).

I think part of this is code structure, as mentioned above:
have a resumer call a function `ingestWithRetry`; that function should be simple:
```
func ingestWithRetry() {
   for retry ... {
     err := ingest()
     // Handle error appropriately
     if retryableError {
        // reload job; log error, etc
     } else {
       // pause the job
    }
```

How we handle errors -- depends.  We could say all errors get retried, or perhaps we retry some of the errors.

gh-casper reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/83310)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @miretskiy, @samiskin, and @stevendanna)

___
*[`pkg/ccl/streamingccl/streamingest/stream_ingestion_job.go` line 139 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/83310#-N5JgMtnAMPrQiJzGBmq-r7-139:-N5gGHReCJJ816Nvp9b_:b-6tzjeh) ([raw file](https://github.com/cockroachdb/cockroach/blob/892ebb1e67e55047933d1ba749d2aed543c91fb5/pkg/ccl/streamingccl/streamingest/stream_ingestion_job.go#L139)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

Are there any cases when the reason why we hit this loop (e.g. retry due to an error) is because the error was with the client? And in such cases we would want to re-create client to re-establish connection?

It feels like this retry loop belongs "on top":
  for retry ... {
     inject()
   }
 
func ingest() {
      client = connect;
       if err { return }
       return distIngest()
  }
</blockquote></details>

Nah, this retry is just make sure the producer job is active/running until we start doing the real ingestion(plan, ingest, cutover). 

This is useful in two scenarios: 
1. when we create the stream ingestion at the first place, the producer job maybe still pending or not adopted yet (it's an adoptable job). 
2. when we resume the stream ingestion through `RESUME JOB id`, we want to make sure the producer job doesn't time out and is still pending, otherwise we will pause on error again.

Yes, you are right about that we should retry the whole ingest as well. I intend to do it in another PR which covers retry in both ingesting stage and clear range stage in the case of cancel. And this work is tracked here. https://github.com/cockroachdb/cockroach/issues/83450
___
*[`pkg/ccl/streamingccl/streamingest/stream_ingestion_job.go` line 140 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/83310#-N5JgMtnAMPrQiJzGBmq-r7-140:-N5gIm3j5FiZLKPIwpVa:bdefm5n) ([raw file](https://github.com/cockroachdb/cockroach/blob/892ebb1e67e55047933d1ba749d2aed543c91fb5/pkg/ccl/streamingccl/streamingest/stream_ingestion_job.go#L140)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

I thought that the ingestion processor was supposed to heartbeat src cluster periodically; I only see this happen here... where do we periodically ping src cluster?
</blockquote></details>

This ping is just to check the producer job status as we need to make sure the producer job is up running either when we first start it or resume from `paused` , Heartbeat is the perfect function to call here.
___
*[`pkg/ccl/streamingccl/streamingest/stream_ingestion_job.go` line 184 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/83310#-N5JgMtnAMPrQiJzGBmq-r7-184:-N5gJRVxCb4kV0Exz7cz:bkz2amq) ([raw file](https://github.com/cockroachdb/cockroach/blob/892ebb1e67e55047933d1ba749d2aed543c91fb5/pkg/ccl/streamingccl/streamingest/stream_ingestion_job.go#L184)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

I don't see in this code how do we ever retry (i.e. hit the `for r := retry.Start(ro); r.Next(); {` line).  It seems that we bubble up every error encountered in this function.

distStreamIngestion for example could return an error for any number of legitimate reasons (node crashed; or perhaps restarted, etc).

I think part of this is code structure, as mentioned above:
have a resumer call a function `ingestWithRetry`; that function should be simple:
```
func ingestWithRetry() {
   for retry ... {
     err := ingest()
     // Handle error appropriately
     if retryableError {
        // reload job; log error, etc
     } else {
       // pause the job
    }
```

How we handle errors -- depends.  We could say all errors get retried, or perhaps we retry some of the errors.
</blockquote></details>

Yes, I agree to this. Retry is a big thing I need to tackle, just didn't want to include it in this PR as it's relatively large and needs its own tests.

https://github.com/cockroachdb/cockroach/issues/83450 tracks this work and its my next step.
___
*[`pkg/jobs/jobspb/jobs.proto` line 114 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/83310#-N5JgMtnAMPrQiJzGBmy-r7-114:-N5gJofK2mHxpy3EOmko:b-bozlwb) ([raw file](https://github.com/cockroachdb/cockroach/blob/892ebb1e67e55047933d1ba749d2aed543c91fb5/pkg/jobs/jobspb/jobs.proto#L114)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

seems like it's a pure code movement? any reasons for this?
</blockquote></details>

Start_time means where we will resume from last ingestion. It's empty when start the ingestion at the first place and it's the last checkpoint when resume ingestion. From the perspective of the whole stream ingestion journey, it's more of a progress information. So I moved it from `StreamIngestionDetails` to `StreamIngestionProgress`.


<!-- Sent from Reviewable.io -->

miretskiy commented:
> Nah, this retry is just make sure the producer job is active/running until we start doing the real ingestion(plan, ingest, cutover).

If that's the purpose, then I suggest you introduce `waitUntilProducerActive` function that does just that.

Even so, I find that your stated purpose of ensuring `the producer job is active/running` is not really met by the
code as I see an error return if `Heartbeat` failed.

gh-casper reviewed (commented):
Sure, moved it into `waitUntilProducerActive `. We should return an error when heartbeat fails. If producer job is not created yet or pending, or other transient status, it won't report error. Error happens when there is connection error or other runtime error on the producer job.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/83310#-:-N5gk_W17fCF9kfxVtlv:bz377ej)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @miretskiy, @samiskin, and @stevendanna)



<!-- Sent from Reviewable.io -->

miretskiy commented:
Also, is it possible to rebase so that already merged PRs do not make
situation more complicated?

On Tue, Jun 28, 2022 at 8:21 PM gh-casper ***@***.***> wrote:

> ***@***.**** commented on this pull request.
>
> Sure, moved it into waitUntilProducerActive . We should return an error
> when heartbeat fails. If producer job is not created yet or pending, or
> other transient status, it won't report error. Error happens when there is
> connection error or other runtime error on the producer job.
>
> *Reviewable
> <https://reviewable.io/reviews/cockroachdb/cockroach/83310#-:-N5gk_W17fCF9kfxVtlv:bz377ej>*
> status: [image: :shipit:] complete! 0 of 0 LGTMs obtained (waiting on
> @miretskiy <https://github.com/miretskiy>, @samiskin
> <https://github.com/samiskin>, and @stevendanna
> <https://github.com/stevendanna>)
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/83310#pullrequestreview-1022604554>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ANA4FVDYUZSEZXNYLMNWE6LVROJI3ANCNFSM5ZW34RIA>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

miretskiy reviewed (changes requested):
I would highly encourage that after addressing comments, this PR is split into independent PRs to simplify review process.

> Reviewed 4 of 10 files at r6, 8 of 15 files at r7, 11 of 14 files at r8, 1 of 1 files at r11, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/83310#-:-N5jckoKEFeYGoCOq0ry:brcfznq)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @gh-casper, @miretskiy, @samiskin, and @stevendanna)

___
*[`pkg/ccl/streamingccl/streamingest/stream_ingestion_frontier_processor.go` line 405 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/83310#-N5jUo4LBLQIBXUg26o_:-N5jUo4LBLQIBXUg26oa:bmeh4qu) ([raw file](https://github.com/cockroachdb/cockroach/blob/7b994112adeacbd953bbf2c41587243e4e5ca86c/pkg/ccl/streamingccl/streamingest/stream_ingestion_frontier_processor.go#L405)):*
> ```Go
> 		if !ok {
> 			err = errors.Errorf("received progress update from unrecognized partition %s", partition)
> 			return true
> ```

I have to say, or even insist, that changes like this deserve to be in separate PR;  The change to
introduce
```
  // Mapping between a source cluster partition ID and its subscribing sql instance ID
  // in the destination cluster. This describes the flow topology of the replication stream.
  map<string, uint32> subscribing_sql_instances = 6 [(gogoproto.nullable) = false,
    (gogoproto.customname) = "SubscribingSQLInstances"];
```
seems innocuous enough.  But it's not -- it's an important change that serializes topology information
on the destination cluster, and ensures that this topology is correct -- for example by returning the error above.  Serialization of state like this raises questions as to the correctness of such state when resuming.


I'm not necessarily saying that this is bad.  But I am saying that such changes *should* be considered carefully.
For example, this is a frontier processor; this progress is initialized when frontier processor has been constructed. Part of this initialization is the initialization of the above subscribing sql instances.
What happens if that set changes?  What happens if say src cluster introduces a new node, and decomissions the old one?  The frontier processor needs to retry the whole flow -- but it's not clear to me what happens with that map.  Is it updated because we retry? Where?  Is it possible that this map gets out of date?

I just have hard time answering these questions and if the review makes it hard for me to answer those questions, perhaps such changes deserve to be reviewed independently and with enough surrounding context.
___
*[`pkg/ccl/streamingccl/streamingest/stream_ingestion_job.go` line 76 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/83310#-N5jYU8pAiDEmhnMKAjA:-N5jYU8pAiDEmhnMKAjB:b4cypwk) ([raw file](https://github.com/cockroachdb/cockroach/blob/7b994112adeacbd953bbf2c41587243e4e5ca86c/pkg/ccl/streamingccl/streamingest/stream_ingestion_job.go#L76)):*
> ```Go
> 		}
> 		lagInfo.SlowestSourcePartitionTimestamp = hlc.MaxTimestamp
> 		lagInfo.FastestSourcePartitionTimestamp = hlc.MinTimestamp
> ```

not sure what these are meant to accomplish... Seems to me that slow/fast is pretty little information
that doesn't actually convey much.  If the whole thing is stuck, both slow and wait will be pretty close;
have you considered just using span.Frontier for this?
___
*[`pkg/ccl/streamingccl/streamingest/stream_ingestion_job.go` line 113 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/83310#-N5jQPVn8JeQePAuVpP-:-N5jQPVn8JeQePAuVpP0:b-2wyerx) ([raw file](https://github.com/cockroachdb/cockroach/blob/7b994112adeacbd953bbf2c41587243e4e5ca86c/pkg/ccl/streamingccl/streamingest/stream_ingestion_job.go#L113)):*
> ```Go
> 
> // Ping the producer job until it is active/running, returns nil when
> // the job is active.
> ```

nit: waits until the job is active?
___
*[`pkg/ccl/streamingccl/streamingest/stream_ingestion_job.go` line 125 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/83310#-N5jQWWl13YivBLbxdxz:-N5jQWWl13YivBLbxdy-:b-lhebvd) ([raw file](https://github.com/cockroachdb/cockroach/blob/7b994112adeacbd953bbf2c41587243e4e5ca86c/pkg/ccl/streamingccl/streamingest/stream_ingestion_job.go#L125)):*
> ```Go
> 		Multiplier:     2,
> 		MaxBackoff:     10 * time.Second,
> 		MaxRetries:     5,
> ```

So, we have 1 + 2 + 4 + 8 + 16 = 31 seconds for the job to become active?
Is that a bit too aggressive?
___
*[`pkg/ccl/streamingccl/streamingest/stream_ingestion_job.go` line 133 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/83310#-N5jRB-zCzUOgMMQ_Kzy:-N5jRB-zCzUOgMMQ_Kzz:b6v83ye) ([raw file](https://github.com/cockroachdb/cockroach/blob/7b994112adeacbd953bbf2c41587243e4e5ca86c/pkg/ccl/streamingccl/streamingest/stream_ingestion_job.go#L133)):*
> ```Go
> 		if err != nil {
> 			return errors.Wrapf(err, "failed to resume ingestion job %d due to producer job error",
> 				ingestionJobID)
> ```

help me understand: we attempt to heartbeat the client.  Connection to source cluster fails; we then immediately return from this function.  What is the point of the retry loop above?
___
*[`pkg/ccl/streamingccl/streamingest/stream_ingestion_job.go` line 232 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/83310#-N5j_4esAh2P93tuF-Lc:-N5j_4esAh2P93tuF-Ld:b-4houeg) ([raw file](https://github.com/cockroachdb/cockroach/blob/7b994112adeacbd953bbf2c41587243e4e5ca86c/pkg/ccl/streamingccl/streamingest/stream_ingestion_job.go#L232)):*
> ```Go
> // Resume is part of the jobs.Resumer interface.
> func (s *streamIngestionResumer) Resume(resumeCtx context.Context, execCtx interface{}) error {
> 	p := execCtx.(sql.JobExecContext)
> ```

nit: while here: `p` is a bit of a remnant from the time when this used to be `Planner`;
mind renaming it to execCtx?
___
*[`pkg/ccl/streamingccl/streamingest/stream_ingestion_job.go` line 241 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/83310#-N5j_lVT64ZF9fhFBA9y:-N5j_lVT64ZF9fhFBA9z:b-xzotkb) ([raw file](https://github.com/cockroachdb/cockroach/blob/7b994112adeacbd953bbf2c41587243e4e5ca86c/pkg/ccl/streamingccl/streamingest/stream_ingestion_job.go#L241)):*
> ```Go
> 		errorMessage := fmt.Sprintf(errorFmt, err)
> 		log.Warningf(resumeCtx, errorFmt, err)
> 		return s.job.PauseRequested(resumeCtx, p.Txn(), func(ctx context.Context,
> ```

What happens on the producer? Should we attempt to best effort pause it?
Do we rely on the fact that the producer should also pause itself?

Please add a comment about this.
___
*[`pkg/ccl/streamingccl/streampb/stream.proto` line 148 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/83310#-N5jbile750N0S-YqvSG:-N5jbile750N0S-YqvSH:bxtcc5a) ([raw file](https://github.com/cockroachdb/cockroach/blob/7b994112adeacbd953bbf2c41587243e4e5ca86c/pkg/ccl/streamingccl/streampb/stream.proto#L148)):*
> ```protobuf
>     // among all partitions.
>     google.protobuf.Duration slowest_fastest_partition_ingestion_lag = 7
>     [(gogoproto.nullable) = false, (gogoproto.stdduration) = true];
> ```

have you considered just storing the span frontier or portion of this frontier
instead of restricting to 2 timestamps/spans?
___
*[`pkg/ccl/streamingccl/streamproducer/stream_lifetime.go` line 252 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/83310#-N5jcM5l6YtFysboI01b:-N5jcM5l6YtFysboI01c:b6bapvf) ([raw file](https://github.com/cockroachdb/cockroach/blob/7b994112adeacbd953bbf2c41587243e4e5ca86c/pkg/ccl/streamingccl/streamproducer/stream_lifetime.go#L252)):*
> ```Go
> 		func(txn *kv.Txn, md jobs.JobMetadata, ju *jobs.JobUpdater) error {
> 			if md.Status == jobs.StatusRunning || md.Status == jobs.StatusPending {
> 				if !ingestionCutover {
> ```

having descriptive status/reason why we are completing stream would improve api; 
it will also allow you to update producer job description and record some metadata -- such as 
when was it completed, why, etc....
___
*[`pkg/streaming/api.go` line 78 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/83310#-N5jSxYMAfVpaAyGIXQC:-N5jSxYMAfVpaAyGIXQD:biuniii) ([raw file](https://github.com/cockroachdb/cockroach/blob/7b994112adeacbd953bbf2c41587243e4e5ca86c/pkg/streaming/api.go#L78)):*
> ```Go
> 	// 'ingestionCutover' indicates whether the stream ingestion has been cutover.
> 	CompleteReplicationStream(
> 		evalCtx *eval.Context, txn *kv.Txn, streamID StreamID, ingestionCutover bool,
> ```

I'm not a fan of this boolean argument; I honestly don't know what it means.
Presumably it's true if it's a "normal" cutover... Why do we care that the source
cluster knows about the reason why it's being canceled? Is it observability?
Does it make sense to have a "reason" as some type of a enum or a struct or a string, or even
an error?
___
*[`pkg/ccl/streamingccl/streamingest/stream_replication_e2e_test.go` line 70 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/83310#-N5jSWkbCoaPvuwtaV9U:-N5jSWkbCoaPvuwtaV9V:bhfxoot) ([raw file](https://github.com/cockroachdb/cockroach/blob/7b994112adeacbd953bbf2c41587243e4e5ca86c/pkg/ccl/streamingccl/streamingest/stream_replication_e2e_test.go#L70)):*
> ```Go
> 	`)
> 	},
> 	srcNumNodes:  1,
> ```

perhaps we should use more than 1


<!-- Sent from Reviewable.io -->

gh-casper reviewed (commented):
My assumption of this whole review process of the two dependent PRs (pause on error and clear range on cancel) was you only review the relevant commits to the PR description. Take this PR for example, only review the top two commits. To review the other commit, please include the comment in the other PR.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/83310#-:-N5safPU3-SWRcqm_nTz:b-el2xb2)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @miretskiy, @samiskin, and @stevendanna)

___
*[`pkg/ccl/streamingccl/streamingest/stream_ingestion_job.go` line 76 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/83310#-N5jYU8pAiDEmhnMKAjA:-N5se_afDGFeIWbL7Y3I:bh3z96c) ([raw file](https://github.com/cockroachdb/cockroach/blob/7b994112adeacbd953bbf2c41587243e4e5ca86c/pkg/ccl/streamingccl/streamingest/stream_ingestion_job.go#L76)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

not sure what these are meant to accomplish... Seems to me that slow/fast is pretty little information
that doesn't actually convey much.  If the whole thing is stuck, both slow and wait will be pretty close;
have you considered just using span.Frontier for this?
</blockquote></details>

See the other comment related to this. Adding a TODO here.
___
*[`pkg/ccl/streamingccl/streamingest/stream_ingestion_job.go` line 125 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/83310#-N5jQWWl13YivBLbxdxz:-N5vjaNx3ortr89nNz2o:bye80m0) ([raw file](https://github.com/cockroachdb/cockroach/blob/7b994112adeacbd953bbf2c41587243e4e5ca86c/pkg/ccl/streamingccl/streamingest/stream_ingestion_job.go#L125)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

So, we have 1 + 2 + 4 + 8 + 16 = 31 seconds for the job to become active?
Is that a bit too aggressive?
</blockquote></details>

Changed it to 1 + 2 + 4 + 5 = 11 seconds.
___
*[`pkg/ccl/streamingccl/streamingest/stream_ingestion_job.go` line 133 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/83310#-N5jRB-zCzUOgMMQ_Kzy:-N5vjnSJ3HgGikg6UvNc:bf00t9e) ([raw file](https://github.com/cockroachdb/cockroach/blob/7b994112adeacbd953bbf2c41587243e4e5ca86c/pkg/ccl/streamingccl/streamingest/stream_ingestion_job.go#L133)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

help me understand: we attempt to heartbeat the client.  Connection to source cluster fails; we then immediately return from this function.  What is the point of the retry loop above?
</blockquote></details>

The retry loop is to deal with transient producer job status like `pending`, `cancel-requested`, `reverting`. Connection failure will make the ingestion error out and pause, instead of retrying here. Users will fix that and resume job manually.
___
*[`pkg/ccl/streamingccl/streamingest/stream_ingestion_job.go` line 241 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/83310#-N5j_lVT64ZF9fhFBA9y:-N5uu9j0Ck-10n-YjJN1:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/7b994112adeacbd953bbf2c41587243e4e5ca86c/pkg/ccl/streamingccl/streamingest/stream_ingestion_job.go#L241)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

What happens on the producer? Should we attempt to best effort pause it?
Do we rely on the fact that the producer should also pause itself?

Please add a comment about this.
</blockquote></details>

Done.



_Code quote:_
```Go
PauseRequested
```
___
*[`pkg/ccl/streamingccl/streamproducer/stream_lifetime.go` line 252 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/83310#-N5jcM5l6YtFysboI01b:-N5vepH_3-nCQSCeTKvb:b-nor1y7) ([raw file](https://github.com/cockroachdb/cockroach/blob/7b994112adeacbd953bbf2c41587243e4e5ca86c/pkg/ccl/streamingccl/streamproducer/stream_lifetime.go#L252)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

having descriptive status/reason why we are completing stream would improve api; 
it will also allow you to update producer job description and record some metadata -- such as 
when was it completed, why, etc....
</blockquote></details>

Made several changes:
Replace `ingestionCutover` with `successfulIngestion` as it makes more sense. Added updating running status when we complete the replication stream.
___
*[`pkg/streaming/api.go` line 78 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/83310#-N5jSxYMAfVpaAyGIXQC:-N5vg3Qn8kBUxh_tIkZR:b7dg7kp) ([raw file](https://github.com/cockroachdb/cockroach/blob/7b994112adeacbd953bbf2c41587243e4e5ca86c/pkg/streaming/api.go#L78)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

I'm not a fan of this boolean argument; I honestly don't know what it means.
Presumably it's true if it's a "normal" cutover... Why do we care that the source
cluster knows about the reason why it's being canceled? Is it observability?
Does it make sense to have a "reason" as some type of a enum or a struct or a string, or even
an error?
</blockquote></details>

It is not the reason we care about; we just need to tell whether ingestion finished successfully so that the producer job can be `succeeded` or `canceled`, so I now replaced `ingestionCutover` with `successfulIngestion`.
___
*[`pkg/ccl/streamingccl/streamingest/stream_replication_e2e_test.go` line 70 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/83310#-N5jSWkbCoaPvuwtaV9U:-N5sa8jHDjk7LwieUV99:by6znrz) ([raw file](https://github.com/cockroachdb/cockroach/blob/7b994112adeacbd953bbf2c41587243e4e5ca86c/pkg/ccl/streamingccl/streamingest/stream_replication_e2e_test.go#L70)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

perhaps we should use more than 1
</blockquote></details>

Using more than 1 doesn't make any difference now as we don't have more than one partition. There is ongoing work to test multiple partitions. I added a TODO item before in the test file, don't think I need to add a TODO here again.
___
*[`pkg/ccl/streamingccl/streamingest/stream_ingestion_frontier_processor.go` line 405 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/83310#-N5jUo4LBLQIBXUg26o_:-N5sbzp2DEPUfkejYbfv:bvr1qa) ([raw file](https://github.com/cockroachdb/cockroach/blob/7b994112adeacbd953bbf2c41587243e4e5ca86c/pkg/ccl/streamingccl/streamingest/stream_ingestion_frontier_processor.go#L405)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

I have to say, or even insist, that changes like this deserve to be in separate PR;  The change to
introduce
```
  // Mapping between a source cluster partition ID and its subscribing sql instance ID
  // in the destination cluster. This describes the flow topology of the replication stream.
  map<string, uint32> subscribing_sql_instances = 6 [(gogoproto.nullable) = false,
    (gogoproto.customname) = "SubscribingSQLInstances"];
```
seems innocuous enough.  But it's not -- it's an important change that serializes topology information
on the destination cluster, and ensures that this topology is correct -- for example by returning the error above.  Serialization of state like this raises questions as to the correctness of such state when resuming.


I'm not necessarily saying that this is bad.  But I am saying that such changes *should* be considered carefully.
For example, this is a frontier processor; this progress is initialized when frontier processor has been constructed. Part of this initialization is the initialization of the above subscribing sql instances.
What happens if that set changes?  What happens if say src cluster introduces a new node, and decomissions the old one?  The frontier processor needs to retry the whole flow -- but it's not clear to me what happens with that map.  Is it updated because we retry? Where?  Is it possible that this map gets out of date?

I just have hard time answering these questions and if the review makes it hard for me to answer those questions, perhaps such changes deserve to be reviewed independently and with enough surrounding context.
</blockquote></details>

It was in the separate PR (stream_ingestion_stats) and this thing already got checked in. 

This map is populated when we create a new flow. We will create a new flow (re-plan) when we will retry. In the future, we will also re-plan on the go if we don't retry, for example, src cluster topology changes and there are so many non-local rangefeeds or dest cluster instances expand or shrink.
___
*[`pkg/ccl/streamingccl/streampb/stream.proto` line 148 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/83310#-N5jbile750N0S-YqvSG:-N5v8DQW1psF2mVWdhLn:b-r1indo) ([raw file](https://github.com/cockroachdb/cockroach/blob/7b994112adeacbd953bbf2c41587243e4e5ca86c/pkg/ccl/streamingccl/streampb/stream.proto#L148)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

have you considered just storing the span frontier or portion of this frontier
instead of restricting to 2 timestamps/spans?
</blockquote></details>

Yes, have considered this and intended to track them once we switch from tracking partition id as frontier to tracking actual spans as frontier in the frontier_processor in the near future. Added a TODO here.


<!-- Sent from Reviewable.io -->

gh-casper reviewed (commented):
But now that we have all the comments in this PR, we just keep them here (no need to comment on the pause-on-erorr PR). The stream-ingestion-stats PR get merged in, no need to worry. I'll merge pause-on-error PR separately once this PR is approved.

Next time, I won't create separate PRs that have dependency again. It just cause too much confusion given that Github doesn't support it well. Also it takes too much time for me to do the syncing and rebasing.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/83310#-:-N5vpHqZ9s-PiohCURKB:bnoe31o)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @erikgrinaker, @miretskiy, @samiskin, and @stevendanna)



<!-- Sent from Reviewable.io -->

miretskiy commented:
>  It just cause too much confusion given that Github doesn't support it well. Also it takes too much time for me to do the syncing and rebasing.

I understand that this is sometimes inconvenient; but you should optimize for the reader -- i.e the person reviewing the code.

gh-casper committed:
streamingccl: expand crdb_internal.complete_stream_replication
to support canceling a producer job

Previously we can only make a producer job normally finish
through complete_stream_replication after ingestion is cutover.
Now we expand complete_stream_replication to take 2nd argument
'successfulIngestion' to support complete replication stream
even when cutover doesn't happen, e.g., ingestion gets canceled
and a revert happens, this cancels the producer job.

Release note (sql change): expand crdb_internal.complete_stream_replication
to take successfulIngestion argument, which indicates if this
stream ingestion finished successfully.

gh-casper reviewed (commented):
Done, Please review.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/83310#-:-N6flSWi1wUTQfNC2SlM:b-5rl50l)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @erikgrinaker, @miretskiy, @samiskin, and @stevendanna)



<!-- Sent from Reviewable.io -->

gh-casper reviewed (commented):
Did some investigation on the bytes-left-after-clearrange bug. In the test, I shut downs the tenant SQL servers before canceling the ingestion. This is not ideal, but we may have a follwup PR that disallow any process to touch this tenant's key range after the ingestion job gets canceled unless user creates another ingestion.

Please review again. @miretskiy @stevendanna 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/83310#-:-N6oTXw8C3NnJoWElaoa:b-ybgaj4)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @erikgrinaker, @miretskiy, @samiskin, and @stevendanna)



<!-- Sent from Reviewable.io -->

gh-casper committed:
streamingccl: support canceling ingestion job

This commit supports the intended behavior when canceling
an ingestion job: canceling the producer job which releases the
protected timestamp for the source tenant. The
destination tenant key ranges are left for
the user to clean up with crdb_internal.destroy_tenant

Release note: None

stevendanna commented on pkg/ccl/streamingccl/streamingest/stream_ingestion_job.go:
```diff
@@ -320,6 +339,24 @@ func activateTenant(ctx context.Context, execCtx interface{}, newTenantID roachp
 	})
 }
 
+func (s *streamIngestionResumer) cancelProducerJob(
```
We don't have to do it in this PR. But I wonder if this should be configurable. Since this will release the protected timestamp on the producer side.

stevendanna commented on pkg/ccl/streamingccl/streamingest/stream_ingestion_job.go:
```diff
@@ -329,7 +366,38 @@ func activateTenant(ctx context.Context, execCtx interface{}, newTenantID roachp
+		return err
+	}
+
+	if tr.State == descpb.TenantInfo_ACTIVE {
+		return errors.Errorf("error in cancel stream ingestion: cannot delete an active tenant %s",
+			details.NewTenantID)
+	}
+
+	// TODO(casper): deal with very large tenant.
+	tenantInfo := descpb.TenantInfo{
+		ID:    p.GetStreamIngestion().NewTenantID.ToUint64(),
+		State: descpb.TenantInfo_DROP,
+	}
+	// GC the tenant's data as well as the system tenant's knowledge about this tenant.
+	if err = sql.GCTenantSync(ctx, jobExecCtx.ExecCfg(), &tenantInfo); err != nil {
+		log.Warningf(ctx,
+			"encountered error when clearing tenant key ranges for ingestion job %d", s.job.ID())
+		return err
+	}
```
Per recent discussions, I think for now we may want to avoid doing this altogether and have a second command: either just `SELECT crdb_internal.destroy_tenant(ID, true)` or a `DROP TENANT` command that we create.

stevendanna commented on pkg/ccl/streamingccl/streamproducer/producer_job.go:
```diff
@@ -84,12 +100,22 @@ func (p *producerJobResumer) Resume(ctx context.Context, execCtx interface{}) er
-			if isTimedOut(j) {
-				return errors.Errorf("replication stream %d timed out", p.job.ID())
+
+			prog := j.Progress()
```
Should we also be checking for this condition at startup to avoid having to wait for one liveness tracking interval?

stevendanna commented on pkg/ccl/streamingccl/streamproducer/producer_job.go:
```diff
@@ -84,12 +100,22 @@ func (p *producerJobResumer) Resume(ctx context.Context, execCtx interface{}) er
+				return p.releaseProtectedTimestamp(ctx, execCfg)
+			case jobspb.StreamReplicationProgress_FINISHED_UNSUCCESSFULLY:
+				return j.Update(ctx, nil, func(txn *kv.Txn, md jobs.JobMetadata, ju *jobs.JobUpdater) error {
+					ju.UpdateStatus(jobs.StatusCancelRequested)
```
Fine for now, but lets think through whether we want to release the protected timestamp in the case of it finishing unnecessfully.

stevendanna reviewed (commented):
Thanks for pushing this forward. As we discussed, I think the only must-have is removing the cleanup on cancellation for now.

gh-casper reviewed (commented):
Was trying to add logic to wait a while after the last ingested time using `TrailingMetaCallback`, but it turns out it's non-trivial work: we have to let the frontier processor to collect trailing `ProducerMetadata`s from all ingestion processors, this needs some modification to the code. Will file another issue for this.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/83310#-:-N7JRy1Q3PuzB46b8L5Y:b-m46yij)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @erikgrinaker, @miretskiy, @samiskin, and @stevendanna)

___
*[`pkg/ccl/streamingccl/streamingest/stream_ingestion_job.go` line 342 at r16](https://reviewable.io/reviews/cockroachdb/cockroach/83310#-N5JgMtnAMPrQiJzGBmq-r16-342:-N7JSjud-VHWj6XlY8E7:b94bp3q) ([raw file](https://github.com/cockroachdb/cockroach/blob/31128b2a536c13508f7f7a04ecb2ff9dc0836eac/pkg/ccl/streamingccl/streamingest/stream_ingestion_job.go#L342)):*
<details><summary><i>Previously, stevendanna (Steven Danna) wrote…</i></summary><blockquote>

We don't have to do it in this PR. But I wonder if this should be configurable. Since this will release the protected timestamp on the producer side.
</blockquote></details>

As mentioned in other comment, we can drop the producer job protected timestamp once we enter `cancel-requested` state.
___
*[`pkg/ccl/streamingccl/streamingest/stream_ingestion_job.go` line 400 at r16](https://reviewable.io/reviews/cockroachdb/cockroach/83310#-N5JgMtnAMPrQiJzGBmq-r16-400:-N7JScdECl5ghEf057X0:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/31128b2a536c13508f7f7a04ecb2ff9dc0836eac/pkg/ccl/streamingccl/streamingest/stream_ingestion_job.go#L400)):*
<details><summary><i>Previously, stevendanna (Steven Danna) wrote…</i></summary><blockquote>

Per recent discussions, I think for now we may want to avoid doing this altogether and have a second command: either just `SELECT crdb_internal.destroy_tenant(ID, true)` or a `DROP TENANT` command that we create.
</blockquote></details>

Done.
___
*[`pkg/ccl/streamingccl/streamproducer/producer_job.go` line 104 at r16](https://reviewable.io/reviews/cockroachdb/cockroach/83310#-N5JgMtnAMPrQiJzGBmv-r16-104:-N7JREgI-nRq5JId-GZs:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/31128b2a536c13508f7f7a04ecb2ff9dc0836eac/pkg/ccl/streamingccl/streamproducer/producer_job.go#L104)):*
<details><summary><i>Previously, stevendanna (Steven Danna) wrote…</i></summary><blockquote>

Should we also be checking for this condition at startup to avoid having to wait for one liveness tracking interval?
</blockquote></details>

Done.



_Code quote:_
```Go
			prog := j.Progress()
```
___
*[`pkg/ccl/streamingccl/streamproducer/producer_job.go` line 110 at r16](https://reviewable.io/reviews/cockroachdb/cockroach/83310#-N5JgMtnAMPrQiJzGBmv-r16-110:-N7JRFl9-gkZZGcQ-rVh:b-uymu90) ([raw file](https://github.com/cockroachdb/cockroach/blob/31128b2a536c13508f7f7a04ecb2ff9dc0836eac/pkg/ccl/streamingccl/streamproducer/producer_job.go#L110)):*
<details><summary><i>Previously, stevendanna (Steven Danna) wrote…</i></summary><blockquote>

Fine for now, but lets think through whether we want to release the protected timestamp in the case of it finishing unnecessfully.
</blockquote></details>

If we decide to cancel and once the job enters `cancel-requested` state, we won't be able to utilize the producer job's protected timestamp anymore: we cannot unpause a cancel-requested job unless we support restoring an existing tenant with deprecated `StartFrom`. Either way, we have to clean up this producer job.


<!-- Sent from Reviewable.io -->

miretskiy reviewed (approved):
> Reviewed 1 of 20 files at r13, 2 of 27 files at r15, 1 of 11 files at r17, 1 of 4 files at r18.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/83310)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @erikgrinaker, @miretskiy, @samiskin, and @stevendanna)



<!-- Sent from Reviewable.io -->

gh-casper commented:
bors r+
